nucliadb/src/nucliadb/ingest/orm/processor/data_augmentation.py

```diff
@@ -0,0 +1,159 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+import logging
+from dataclasses import dataclass, field
+from typing import Optional
+
+from nucliadb.ingest.orm.resource import Resource
+from nucliadb.ingest.processing import ProcessingEngine, PushPayload, Source
+from nucliadb_models.text import PushTextFormat, Text
+from nucliadb_protos import resources_pb2, writer_pb2
+from nucliadb_protos.resources_pb2 import FieldType
+from nucliadb_utils.utilities import Utility, get_partitioning, get_utility
+
+logger = logging.getLogger("ingest-processor")
+
+
+@dataclass
+class GeneratedFields:
+    texts: list[str] = field(default_factory=list)
+    links: list[str] = field(default_factory=list)
+    files: list[str] = field(default_factory=list)
+    conversations: list[str] = field(default_factory=list)
+
+    def __len__(self):
+        return len(self.texts) + len(self.links) + len(self.files) + len(self.conversations)
+
+
+async def has_generated_fields(bm: writer_pb2.BrokerMessage, resource: Resource) -> bool:
+    generated_fields = await _get_generated_fields(bm, resource)
+    return len(generated_fields) > 0
+
+
+async def _get_generated_fields(bm: writer_pb2.BrokerMessage, resource: Resource) -> GeneratedFields:
+    """Processing can send messages with generated fields. Those can be
+    generated with a data augmentation task and, as learning can't queue it to
+    process, nucliadb is responsible to send the generated field to process (and
+    ingest the processed thing later).
+
+    Given a broker message and a resource, this function returns the list of
+    generated fields, that can be empty.
+
+    """
+    generated_fields = GeneratedFields()
+
+    # only messages from processor can add generated fields
+    if bm.source != writer_pb2.BrokerMessage.MessageSource.PROCESSOR:
+        return generated_fields
+
+    # search all fields
+
+    for field_id in bm.texts:
+        if not resource.has_field(FieldType.TEXT, field_id):
+            generated_fields.texts.append(field_id)
+
+    for field_id in bm.links:
+        if not resource.has_field(FieldType.LINK, field_id):
+            generated_fields.links.append(field_id)
+
+    for field_id in bm.files:
+        if not resource.has_field(FieldType.FILE, field_id):
+            generated_fields.files.append(field_id)
+
+    for field_id in bm.conversations:
+        if not resource.has_field(FieldType.CONVERSATION, field_id):
+            generated_fields.conversations.append(field_id)
+
+    return generated_fields
+
+
+async def send_generated_fields_to_process(
+    kbid: str,
+    resource: Resource,
+    bm: writer_pb2.BrokerMessage,
+):
+    generated_fields = await _get_generated_fields(bm, resource)
```

Comment:
We are calling the `_get_generated_fields` function twice (first to check if there are any, second here).

This would be fine, except that `_get_generated_fields` calls `has_field` multiple times. `has_field` is not cached, each call is a call to the database. So we end up with 2*number_of_fields redundant calls to the database.

So, two things to see here:
1. In `_get_generated_fields` avoid calling `has_field` multiple times. Instead, call `get_all_field_ids` and compare if the incoming field is in that list.
2. Avoid calling `_get_generated_fields`. The usual approaches are:
  a) Moving the check inside this function
  b) Making this function take the list of generated fields by parameter.

Reply:
Solved with option 2

Reply:
Still missing the problem where the function calls `has_field` repeatedly, and each call is a query.